src/ray/util/event.cc

```diff
@@ -15,6 +15,93 @@
+/// LogEventReporter
+///
+LogEventReporter::LogEventReporter(rpc::Event_SourceType source_type,
+                                   std::string &log_dir, bool force_flush,
+                                   int rotate_max_file_size, int rotate_max_file_num)
+    : log_dir_(log_dir),
+      force_flush_(force_flush),
+      rotate_max_file_size_(rotate_max_file_size),
+      rotate_max_file_num_(rotate_max_file_num) {
+  RAY_CHECK(log_dir_ != "");
+  if (log_dir_.back() != '/') {
+    log_dir_ += '/';
+  }
+  boost::filesystem::create_directories(log_dir_);
+  file_name_ = "event_" + Event_SourceType_Name(source_type);
+  if (source_type == rpc::Event_SourceType::Event_SourceType_CORE_WORKER ||
+      source_type == rpc::Event_SourceType::Event_SourceType_COMMON) {
+    file_name_ += "_" + std::to_string(getpid());
+  }
+  file_name_ += ".log";
+
+  std::string log_sink_key = GetReporterKey() + log_dir_ + file_name_;
+  log_sink_ = spdlog::get(log_sink_key);
+  if (log_sink_ == nullptr) {
+    log_sink_ =
+        spdlog::rotating_logger_mt(log_sink_key, log_dir_ + file_name_,
+                                   1048576 * rotate_max_file_size_, rotate_max_file_num_);
+  }
+  log_sink_->set_pattern("%v");
+}
+
+LogEventReporter::~LogEventReporter() { Flush(); }
+
+void LogEventReporter::Flush() { log_sink_->flush(); }
+
+std::string LogEventReporter::EventToString(const rpc::Event &event) {
+  std::stringstream result;
+
+  boost::property_tree::ptree pt;
+
+  auto time_stamp = event.timestamp();
+  std::stringstream time_stamp_buffer;
+  char time_buffer[30];
+  time_t epoch_time_as_time_t = time_stamp / 1000000;
+
+  struct tm *dt = localtime(&epoch_time_as_time_t);
+  strftime(time_buffer, sizeof(time_buffer), "%Y-%m-%d %H:%M:%S.", dt);
+
+  time_stamp_buffer << std::string(time_buffer) << std::setw(6) << std::setfill('0')
+                    << time_stamp % 1000000;
+
+  pt.put("time_stamp", time_stamp_buffer.str());
+  pt.put("severity", Event_Severity_Name(event.severity()));
+  pt.put("label", event.label());
+  pt.put("event_id", event.event_id());
+  pt.put("source_type", Event_SourceType_Name(event.source_type()));
+  pt.put("host_name", event.source_hostname());
+  pt.put("pid", std::to_string(event.source_pid()));
+  pt.put("message", event.message());
+
+  boost::property_tree::ptree pt_child;
+  for (auto &ele : event.custom_fields()) {
+    pt_child.put(ele.first, ele.second);
+  }
+
+  pt.add_child("custom_fields", pt_child);
+
+  std::stringstream ss;
+  boost::property_tree::json_parser::write_json(ss, pt, false);
+
+  return ss.str();
+}
+
+void LogEventReporter::Report(const rpc::Event &event) {
+  RAY_CHECK(Event_SourceType_IsValid(event.source_type()));
+  RAY_CHECK(Event_Severity_IsValid(event.severity()));
+  std::string result = EventToString(event);
+  // Pop the last character from the result string because it is breakline '\n'.
+  result.pop_back();
+
+  log_sink_->info(result);
+  if (force_flush_) {
+    Flush();
```

Comment:
When do we need this attribute? Is it for testing? Can you clearly specify this attribute is for testing?

Reply:
force_flush_ is used to clear the file buffer. When fwrite function is called, sometimes the information is not actually written to the file, but stored in the OS file buffer. The flush will be triggered when it exceeds certain buffer size.

Reply:
I understand when we need flush! But isn’t this triggered automatically by spdlog if buffers are filled up? I just don’t know when we need to set this attribute true except test scenario.

Reply:
When the system receives SIGTERM, the log reporter's destructor may not be triggered. Some of the last event information may not be sent successfully. But if it is force_flush mode, because it will be flushed when RAY_EVENT is called, so we don’t worry about this problem.